lesson_14/task_2.py

```diff
@@ -0,0 +1,106 @@
+# Describe the Train object. The class must contain fields and a method
+# for adding wagons (it is necessary to add objects, instances of the wagon class). +
+
+# Describe the class Wagon together with the train. The Wagon must contain a list of
+# passengers and allow adding passengers. In the Wagon can be 10 passengers for example. ++
+
+# When using the len function on a Wagon, I want to see the number of passengers. + When using len on a train,
+# I want to see a list of Wagons without a locomotive. +
+
+# Each wagon must have a number. +
+# When printing a wagon to the console, I want to see the following "[n]" where n is the number of the wagon. +
+
+class Train:
+
+    def __init__(self):
+        self.__list_of_wagons = ['Locomotive']
+        self.__num_wagons = 1  # represents locomotive
+
+    def add_wagon(self, wagon):
+        # wagon - is an instance of Wagon class
+        if not isinstance(wagon, Wagon):
+            raise TypeError(f'The "wagon" should be an instance of Wagon class ')
+        # prevent from adding already existing wagon number to the train
+        if wagon in self.__list_of_wagons:
+            raise ValueError(f'Wagon {wagon} already exists in the train')
+        self.__list_of_wagons.append(wagon)
+        self.__num_wagons += 1
+
+    def __len__(self):
+        return self.__num_wagons - 1
+
+    def __str__(self):
+        return f'Train has {[str(wagon) for wagon in self.__list_of_wagons[1:]]} wagons'
+
+
+class Wagon:
+
+    __wagons = []
+    __passengers_in_wagon = {}
```

Comment:
why is this var outside init?

Reply:
because I wanted to try to use __setattr__ method, and in this method I add logic to check if wagon number in list of wagons or not. And if I add __wagons = [] to init, I either need to make it required attribute, or __setattr__ will not see __wagons = [] and code will rase an attribute error. So I decided to use it outside of constructor